packages/autocomplete-plugin-recent-searches/src/usecases/localStorage/createSearchableLocalStorageStore.ts

```diff
@@ -0,0 +1,37 @@
+import { createStore } from '../../createStore';
+import { RecentSearchesItem } from '../../types/RecentSearchesItem';
+
+import { createLocalStorage } from './createLocalStorage';
+
+export type CreateSearchableLocalStorageRecentSearchesStoreProps = {
+  key: string;
+  limit: number;
+};
+
+export function createSearchableLocalStorageStore<
+  TItem extends RecentSearchesItem
+>({ key, limit }: CreateSearchableLocalStorageRecentSearchesStoreProps) {
+  const storage = createLocalStorage<TItem>({ key });
+  const store = createStore<TItem>({
+    getAll(query) {
+      if (query) {
+        return storage
+          .getItem()
+          .filter((item) =>
+            item.query.toLowerCase().startsWith(query.toLowerCase())
+          )
+          .splice(0, limit);
```

Comment:
what if this can be overridden from the plugin api? that way you could plug more advanced search methods

Reply:
Yeah that makes sense—something like:

```ts
const recentSearches = createSearchableLocalStorageRecentSearchesPlugin({
  key: 'recent',
  search({ query, items }) {
    return items.filter((x) => x.includes(query));
  },
});
```

It would even allow to export a single `createLocalStorageRecentSearchesPlugin` function.

If we agree on this, I can either update this PR or create a stacked PR.

Reply:
exactly, that's what I was thinking as well. We could export localStorageRecentSearchesPlugin and simpleLocalSearch or something like that.

To be pluggable with more advanced libraries you might want to change the backend to them so they can index though (Algolia, but also fuse, lunr, flexsearch)

Reply:
I actually wonder if the api might not better be a merge between the backend & the searching:

```js
const recentSearches = createRecentSearchesPlugin({
  key: 'recent',
  setItems({ items }) {
    // default something with localstorage
  },
  getItems({ query }) {
    // could be all items or already filtered
  },
  search({ query, items }) {
    // so you can override the search, but leave localstorage as is
  }
});
```

then you can provide a default getItems / setItems for localstorage & search for simple search, but contacting a different backend would be easy

Reply:
Addressed in 1d7e6bf2a8e7b9a752f977a21df865bbb028566d.